### PR TITLE
New branch

### PR DESCRIPTION
**Pull request by staceycoils**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,7 @@
 # Northcoders News API
+
+## Envirmonment Setup
+
+To ensure envirmonent variables are pathed correctly, create two files in the root directory: .en.test and .env.development.
```
Comment (DominicFH) [inline README.md]: Teeny weeny typo with '.en.test' !

File: __tests__/app.test.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        .get("/api/topics")
+        .expect(200)
+    });
+    test("returns an array of objects with properties 'slug', 'description", () => {
+        return request(app)
+            .get("/api/topics")
+            .expect(200)
+        .then((response) => {
+            expect(response.body).toHaveLength(3);
+            response.body.forEach((cell) => {
```
Comment (DominicFH) [inline __tests__/app.test.js]: I'd maybe rename 'cell' to be 'topic' just to be more clear about what exactly is within this array 😄

File: app.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    getTopics, 
+    getArticles
+} = require('./controllers/controllers.js');
+
+const app = express();
+app.use(express.json());
+
+app.get("/api/topics", getTopics);
+
+app.get(/^\/api\/articles/, getArticles);
```
Comment (emilyb93) [inline app.js]: No need to use regex here as express already does this if we just pass a string of the endpoint, much like you have done for "/api/topics" above

File: controllers/controllers.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+function getArticles(req,res,next) {
+    if (req.path.endsWith("articles")) {
+        fetchArticles()
+        .then((data) => {
+            res.status(200).send(data)
+        })
+    } else {
+        let index = ((req.path).slice(14))
```
Comment (emilyb93) [inline controllers/controllers.js]: This technically works but i want you to look into req.params to get the parametric endpoints and later on, look at req.query to get any queries too

File: models/models.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+const db = require('../db/connection.js')
+
```
Comment (emilyb93) [inline models/models.js]: General note: always close your sql strings with a ;

File: models/models.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    return db.query("SELECT * FROM articles")
+        .then((data) => {
+            return data.rows
+        })
+}
+
+function fetchArticle(num) {
+    return db.query("SELECT * FROM articles")
+        .then((data) => {
+            if (!data.rows[num]) return Promise.reject();
```
Comment (emilyb93) [inline models/models.js]: This needs changing; if i request article number 3, rows should still have a length of one and therefore rows[3] wouldn't exist but it's still a valid request. Its best to check if there is anything at rows[0] or check the length of rows instead

**Review by emilyb93 — CHANGES_REQUESTED** [review]
Left some notes :)

**Review by DominicFH — CHANGES_REQUESTED** [review]
There are some comments left from myself and Emily but on top of those I'd also refactor your models and controllers to have separate files for articles controllers and models as well as topics controllers and models 😄

File: __tests__/app.test.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+const app = require("../app.js");
+const seed = require('../db/seeds/seed')
+const data = require('../db/data/test-data')
+const db = require('../db/connection.js')
+
+beforeEach(() => seed(data))
+
+afterAll(() => db.end());
+
+describe("GET /*", () => {
```
Comment (maddierosecodes) [inline __tests__/app.test.js:11]: I would describe this as not-a-path rather than '/*' which implies all paths

File: __tests__/app.test.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+      return request(app)
+        .get("/api/articles/3")
+        .expect(200)
+    });
+    test("returns a specified object from the database", () => {
+        return request(app)
+            .get("/api/articles/3")
+            .expect(200)
+        .then((response) => {
+            expect(Object.keys(response.body)).toHaveLength(7);
```
Comment (maddierosecodes) [inline __tests__/app.test.js:56]: What if we were to add extra properties in the future?

File: __tests__/app.test.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            expect(response.body).toEqual({
+                article_id: 3,
+                title: "Eight pug gifs that remind me of mitch",
+                topic: "mitch",
+                author: "icellusedkars",
+                body: "some gifs",
+                created_at: '2020-11-03T09:12:00.000Z',
+                votes: 0,
```
Comment (maddierosecodes) [inline __tests__/app.test.js:64]: we can use expect.objectContaining() to ensure the test doesn't break if more properties are added in the future

File: models/models_articles.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+        .then((data) => {
+            return data.rows
+        })
+}
+
+function fetchArticle(num) {
+    return db.query(`SELECT * FROM articles
+                    WHERE article_id = ${num};`)
+        .then((data) => {
+            if (!data.rows[0]) return Promise.reject();
```
Comment (maddierosecodes) [inline models/models_articles.js:14]: Have a look at the notes on advanced error handling for how to handle cases where we do want an empty array back

**Review by maddierosecodes — APPROVED** [review]
I'm approving this branch now with pending changes so that you can merge the branch with main. Please fix these few changes on main and push them to main. Then, please create a new branch for the next ticket, and keep creating new branches for every different ticket you work on.

**Pull request merged**